controller/array_action/array_mediator_ds8k.py

```diff
@@ -0,0 +1,412 @@
+HOST_VOLUME_MAPPING_VOLUME_ID = 'volume_id'
+HOST_VOLUME_MAPPING_LUN_ID = 'lunid'
+HOST_PORTS = 'host_ports_briefs'
+HOST_ADDRESS_DISCOVERY = 'addrdiscovery'
+
+# host port
+HOST_PORT_WWPN = 'wwpn'
+HOST_PORT_HOST_ID = 'host'
+HOST_PORT_IOPORTS = 'login_ports'
+HOST_PORT_STATE = 'state'
+HOST_PORT_TYPE = 'hosttype'
+
+# host mapping
+HOST_MAPPING_LUN_ID = 'lunid'
+HOST_MAPPING_VOLUME_ID = 'volume'
+
+# user
+USER_NAME = 'name'
+USER_STATE = 'state'
+USER_GROUP = 'group'
+
+# lss
+LSS_ID = 'id'
+LSS_TYPE = 'type'
+LSS_GROUP = 'group'
+
+# io enclosure
+IOENCLOSURE_ID = 'id'
+IOENCLOSURE_NAME = 'name'
+IOENCLOSURE_STATE = 'state'
+
+# node
+NODE_ID = 'id'
+NODE_STATUS = 'state'
+
+# marray
+MARRAY_ID = 'id'
+MARRAY_DISK_CLASS = 'disk_class'
+MARRAY_POOL_ID = 'pool'
+
+# response keys
+RES_STATUS = 'status'
+RES_CODE = 'code'
+RES_MSG = 'message'
+
+# response values
+RES_SUCCESS_STAUTS = ('ok', 'successful', )
+
+
+def parse_version(bundle):
+    """
+    Parse ds8k version number form bundle.
+
+    rule is:
+    87.51.34.0 => 7.5.1
+    87.50.21.1 => 7.5.0
+    88.0.151.0 => 8.0.0
+    """
+
+    v1, v2, _ = bundle.split('.', 2)
+    v2 = '{0:0{1}}'.format(int(v2), 2)
+
+    return '.'.join([v1[-1], v2[0], v2[1]])
+
+
+def get_volume_id_from_scsi_identifier(scsi_id):
+    return scsi_id[-4:]
+
+
+def build_kwargs_from_capabilities(capabilities, pool_id, name, size):
+    cli_kwargs = {}
+    size_in_gib = int(math.ceil(convert_size_bytes_to_gib(size)))
+    cli_kwargs.update({
+        'pool_id': pool_id,
+        'capacity_in_gib': size_in_gib,
+        'tp': 'none',
+        'volume_names_list': [shorten_volume_name(name), ],
+    })
+    capability = capabilities.get(config.CAPABILITIES_SPACEEFFICIENCY)
+    if capability:
+        capability = capability.lower()
+        if capability == config.CAPABILITY_THIN:
+            cli_kwargs.update({'tp': 'ese'})
+
+    return cli_kwargs
+
+
+# shorten volume name to 16 characters if it is too long.
+def shorten_volume_name(name):
+    if len(name) <= 16:
+        return name
+    return sha1(name.encode()).hexdigest()[-16:]
+
+
+class DS8KArrayMediator(ArrayMediator):
+    SUPPORTED_FROM_VERSION = '7.5.1'
+
+    @classproperty
+    def array_type(self):
+        return 'DS8K'
+
+    @classproperty
+    def port(self):
+        return 8452
+
+    @classproperty
+    def max_vol_name_length(self):
+        return 16
+
+    @classproperty
+    def max_connections(self):
+        return 20
+
+    @classproperty
+    def minimal_volume_size_in_bytes(self):
+        return convert_size_gib_to_bytes(1)
+
+    @classproperty
+    def max_lun_retries(self):
+        return 10
+
+    def __init__(self, user, password, endpoint):
+        self.user = user
+        self.service_address = \
+            endpoint[0] if isinstance(endpoint, list) else endpoint
+        self.password = password
+        self.client = None
+        self._system_info = None
+
+        self._connect()
+
```

Comment:
sounds good. couldn't find the refactor yet, so I unresolved for now.

Reply:
you can see the example in line 241.

Reply:
I see a call to `get_system_info`, and `get_system_info` still checks `if self._system_info is None:`.
If `self._system_info` is initialized in `__init__` (/`_connect`), then there will be no need to this check.

Reply:
it is just a code style to make sure system_info is valid, because it will be used in several properties. Maybe for some reasons, the _system_info is cleared.

Reply:
if there is no code that clears `_system_info`, then no need to validate it each time.
otherwise, you would have e.g. `if self._system_info is None and self.client is not None:`.
please initialize `self._system_info` during `__init__ `(/`_connect)`, and remove unnecessary checks.

the point of `__init__ ` methods is to ensure the initialized state of the instance (i.e. instance members), so the other methods wouldn't need those checks everywhere.

Reply:
I do initialize the _system_info in _connect by calling property "version", and I will keep this check.

Reply:
This is inefficient, since every property access causes an unnecessary check.
Which of the following examples is more efficient (and shorter)?:
1.
```
class A():
    def __init___(self):
        self.info = Info()
    def get_info(self):
        return self.info
    @property
    def id(self):
        return self.info.id
    @property
    def name(self):
        return self.info.name
    @property
    def version(self):
        return self.info.version
```
2.
```
class B():
    def __init___(self):
        self.info = self.get_info()
    def get_info(self):
        if not self.info:
            self.info = Info()
        return self.info
    @property
    def id(self):
        return self.get_info().id
    @property
    def name(self):
        return self.get_info().name
    @property
    def version(self):
        return self.get_info().version
```

Why do you want to keep this check?

Reply:
ok, updated.